netket/optimizer/jax/_sr_onthefly.py

```diff
@@ -0,0 +1,105 @@
+import jax
+import jax.numpy as jnp
+from functools import partial
+from netket.stats import sum_inplace as _sum_inplace
+from netket.utils import n_nodes
+
+
+# onthefly SR
+# since we cant store O if #samples x #params is too large
+
+
+def O_jvp(x, theta, v, forward_fn):
+    _, res = jax.jvp(lambda p: forward_fn(p, x), (theta,), (v,))
+    return res
+
+
+def O_vjp(x, theta, v, forward_fn, return_vjp_fun=False, vjp_fun=None):
+    if vjp_fun is None:
+        _, vjp_fun = jax.vjp(forward_fn, theta, x)
+    res, _ = vjp_fun(v)
+    if return_vjp_fun:
+        return res, vjp_fun
+    else:
+        return res
+
+
+# calculate \langle O \rangle
+def Obar(samples, theta, forward_fn, **kwargs):
+    # TODO better way to get dtype
+    dtype = forward_fn(theta, samples[:1])[0].dtype
+    v = jax.lax.broadcast(
```

Comment:
can you please add a comment on what broadcast is doing here?

Reply:
Yep. In general if you could better document what exactly is the flow of the code among the several small functions you defined, it would be great.

It's a bit hard to follow.

Reply:
Do you have suggestion on how I could solve the TODO with the dtype here?

Reply:
There is not really a way to avoid that. 
However, please check that the forward pass is being optimised away (I believe it should be, but I'm not 100% sure).
You can check on the generated IR by using `jax.make_jaxpr(Obar)(samples, theta...)`.

If it does not cut out the forward pass, then we will need to pass it as a static arg.